### PR TITLE
Create CMake option `onnxruntime_USE_VCPKG`

### PR DESCRIPTION
**Pull request by luncliff**
### Changes

1. CMake option `onnxruntime_USE_VCPKG`. It will be used in the vcpkg port
   * Unit test may fail because this option leads to a mixture of unexpected external library versions.  
     Especially ONNX, Protobuf, and Flatbuffers version can be different
2. Overhaul of `onnxruntime_external_deps.cmake`
   * Make `FetchContent_Declare` to try `find_package`.  
     See https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html
   * Relocated `FetchContent_Declare` and `FetchContent_MakeAvailable`(or `onnxruntime_fetchcontent_makeavailable`) to closer lines.  
     It was too hard to navigate the entire file to search related sections...
   * Alias `IMPORTED` targets like build targets (e.g. `ONNX::onnx` --> `onnx`)

```cmake
# The script uses `find_package` with the changes.
# In this case, use vcpkg to search dependencies
# See https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html
include(external/onnxruntime_external_deps.cmake)
```

3. Create CMakePresets.json and presets to [run vcpkg in manifest mode](https://learn.microsoft.com/en-us/vcpkg/concepts/manifest-mode)
   * Currently, it's NOT for training build
   * Main triplets are `x64-windows` and `x64-osx`

```pwsh
Push-Location "cmake"
    cmake --preset "x64-windows-vcpkg"
    cmake --build --preset "x64-windows-vcpkg-debug"
Pop-Location
```
```bash
pushd "cmake"
    cmake --preset "x64-osx-vcpkg"
    cmake --build --preset "x64-osx-vcpkg-debug"
popd
```

4. Updated tools/ci_build/build.py
   * `--use_vcpkg` option: it needs `CMAKE_TOOLCHAIN_FILE` with [vcpkg.cmake toolchain script](https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake)
   * `--compile_no_warning_as_error` is recommended because library version differences will cause unexpected compiler warnings

```bash
python ./tools/ci_build/build.py \
    --compile_no_warning_as_error \
    --use_vcpkg \
    --cmake_extra_defines "CMAKE_TOOLCHAIN_FILE:FILEPATH=${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake" \
    --cmake_extra_defines "VCPKG_TARGET_TRIPLET=..."
```

5. Created Job `Vcpkg` for Windows and macOS
   * Show how to setup and use vcpkg.  
     Similar to the CMakePresets.json usage

### Motivation and Context

* Help #7150
* Help https://github.com/microsoft/vcpkg/pull/36850
   * https://github.com/luncliff/vcpkg-registry/pull/212
   * https://github.com/microsoft/vcpkg/pull/39881
* https://github.com/luncliff/vcpkg-registry/pull/215
   * https://github.com/luncliff/vcpkg-registry/pull/216
   * https://github.com/luncliff/vcpkg-registry/pull/227
* https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html
* https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake

### Future Works?

More feature coverage with the vcpkg supported libraries

* CUDA feature support
* Training feature support

### CODE REVIEW

**Comment by snnn** [comment]
I think cmake is a great choice. 
Which cmake triplet this PR is going to support?

Besides, we might decouple the work. You may create a vcpkg config file, then install all the packages via "vcpkg install", then let cmake search installed packages there when running ONNX Runtime's build.py script. Then in theory you don't need to modify any ONNX  Runtime's cmake file.  No matter which way you choose to go, we always greatly appreciate your contribution.

**Comment by luncliff** [comment]
https://github.com/microsoft/onnxruntime/pull/21348/commits/36919b0eb7d3baccc00b4a183ebf36e88dc74959 New GitHub Actions jobs for vcpkg build

https://github.com/microsoft/onnxruntime/pull/21348/commits/6227f5fdc8c1602f5c848378abfec5627d576853 Trying to merge `onnxruntime_vcpkg_deps.cmake` into `onnxruntime_external_deps.cmake` using [`FIND_PACKAGE_ARGS `](https://cmake.org/cmake/help/latest/module/FetchContent.html#command:fetchcontent_declare).
Currently I have some problems with pb.h,pb.cc generation

* https://cmake.org/cmake/help/latest/module/FetchContent.html#command:fetchcontent_declare 
* https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html

File: cmake/vcpkg-configuration.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            "repository": "https://github.com/microsoft/vcpkg",
+            "kind": "git",
+            "packages": [
+                "flatbuffers"
+            ],
+            "baseline": "8161f655436492669adc8eb6f269dc41392ffa70"
+        },
+        {
+            "repository": "https://github.com/luncliff/vcpkg-registry",
```
Comment (snnn) [inline cmake/vcpkg-configuration.json:17]: This is your personal repo?

Comment (luncliff) [reply cmake/vcpkg-configuration.json:17]: Yes. Using onyx 1.16.0 from vcpkg upstream is contained in https://github.com/microsoft/vcpkg/pull/36850
The part can be removed after the PR

**Comment by luncliff** [comment]
Updated note and title.

Dropped CUDA feature support in this PR.  
I felt like making too huge changes here. I would make another PR for other features

**Comment by snnn** [comment]
We probably need to wait for the next absl release.

**Comment by luncliff** [comment]
`absl::demangle_rust` was not intended one. I saw `absl::demangle_internal` in my fork, but not the one in the log. I will check again

**Comment by luncliff** [comment]
> We probably need to wait for the next absl release.

So I confused your message days ago.

Update vcpkg-configuration.json to use later version of `abseil-cpp`.

**Comment by luncliff** [comment]
Commit https://github.com/microsoft/onnxruntime/pull/21348/commits/2889abfe55edbce2701a8b7330df7eb40628a120 use different version of https://gitlab.com/libeigen/eigen

In microsoft/vcpkg, the version is 3.4.0
https://github.com/microsoft/vcpkg/blob/fe1cde61e971d53c9687cf9a46308f8f55da19fa/versions/e-/eigen3.json#L3-L7

In luncliff/vcpkg-registry,  
the version is 2024-01-16, which is a date of https://gitlab.com/libeigen/eigen/-/tree/b2814d53a707f699e5c56f565847e7020654efc2

File: cmake/vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    },
+    {
+      "name": "boost-mp11",
+      "version>=": "1.82.0"
+    },
+    "cpuinfo",
+    "cxxopts",
+    "date",
+    "dlpack",
+    "eigen3",
```
Comment (snnn) [inline cmake/vcpkg.json:21]: Can we set eigen3's version here?

**Comment by luncliff** [comment]
Hi, is there a plan for this PR?
I'm going to restart discussion in vcpkg, so it will be helpful if I know the milestone of the project.

Any feedback will be welcomed. Thanks 😊

**Comment by snnn** [comment]
Sorry I was on vacation last month. Just came back.

File: cmake/vcpkg.json (hunk 2 above)
Comment (snnn) [reply cmake/vcpkg.json:22]: How can we avoid depending on your personal repos?

**Comment by luncliff** [comment]
Probably with the latest vcpkg release. I will check it within days

File: cmake/external/onnxruntime_external_deps.cmake [hunk 3]
```diff
@@ -254,7 +298,16 @@ FetchContent_Declare(
   mp11
   URL ${DEP_URL_mp11}
   URL_HASH SHA1=${DEP_SHA1_mp11}
+  FIND_PACKAGE_ARGS NAMES Boost
 )
+onnxruntime_fetchcontent_makeavailable(mp11)
+if(NOT TARGET Boost::mp11)
+  if(onnxruntime_USE_VCPKG)
```
Comment (snnn) [inline cmake/external/onnxruntime_external_deps.cmake:305]: By default FetchContent will call find_package first. Why this change is needed?

File: cmake/vcpkg-configuration.json [hunk 4]
```diff
@@ -2,18 +2,7 @@
     "default-registry": {
         "kind": "git",
         "repository": "https://github.com/Microsoft/vcpkg",
-        "baseline": "1de2026f28ead93ff1773e6e680387643e914ea1"
+        "baseline": "3508985146f1b1d248c67ead13f8f54be5b4f5da"
```
Comment (snnn) [inline cmake/vcpkg-configuration.json:5]: Thanks!

**Review by snnn — APPROVED** [review]

**Comment by snnn** [comment]
Thank you!

**Comment by luncliff** [comment]
Thank you, @snnn. I will submit another PR if microsoft/vcpkg needs it.
